### ISSUE TITLE
Multiprocessed/multithreaded data loading

### ISSUE THREAD

**Issue by joelvbernier** [HEXRD/hexrdgui#310]
In the course of using the GUI for an 8-detector setup, it has become clear that we desperately need to multiprocess or multithread the load data actions across the individual panels.  Not sure which one is better from a performance point of view, but it would be simple to throw to a process pool as they run independently.

**Comment by psavery** [comment]
I think I have some good work towards this. I have one example that is running much faster due to multiprocessing.

It looks to me that multiprocessing might be better for us than multithreading because for multithreading, the GIL will only allow one thread of python code to execute at a time. The [multiprocessing library](https://docs.python.org/3/library/multiprocessing.html) also makes multiprocessing easy.

However, for multiprocessing, the objects that are passed to the processes need to be pickle-able. It looks like the framecache (.npz files) imageseries is pickle-able, but hdf5 imageseries is not (since it holds onto an h5py object, which is not pickle-able, but we may be able to change that).

@joelvbernier Is there a way we can get the data for the 8 detector setup so we can test with that as well?

**Comment by psavery** [comment]
Actually, since the numpy functions are implemented in C/Fortran, it might also be faster to do multithreading instead of multiprocessing because it avoids the pickling. Maybe the best option is to benchmark.

**Comment by psavery** [comment]
As a side note on this issue as well, I noticed that, for my examples, if you perform dark background subtraction, it takes significantly longer (5-8x longer) to aggregate the images. It looks like this is because, in my examples, the original data is uint16, and after we perform the background subtraction, the output data is float64. This increases the size of our data by 4x.

Can we truncate that back down to a uint16 before aggregating in order to significantly improve performance?

**Comment by donald-e-boyce** [comment]
That's a good question and the issue should be clarified.  

The imageseries.process module (line 57) preserves the datatype on dark subtraction. That made most sense to me.
```
def _subtract_dark(self, img, dark):
    # need to check for values below zero
    # !!! careful, truncation going on here;necessary to promote dtype?
    return np.where(img > dark, img - dark, 0)
```

**Comment by psavery** [comment]
@donald-e-boyce Yeah, so it looks like when we create a dark image using one of the aggregation functions (like `imageseries.stats.median`), the input is `np.uint16`, and the output is `np.float64`, which probably makes sense, because the stats output can be fractions.

Then the dark image gets added to the list of ops, and then when it is passed into `imageseries.process.ProcessedImageSeries`, the input imageseries is `np.uint16`, the dark image is `np.float64`, and the output is `np.float64`.

Should we get the processed image series to keep the same dtype as the input imageseries?

It looks like if I do `np.where(a > b, a - b, 0)` where `a` is `np.uint16` and `b` is `np.float64`, the output is `np.float64`.

**Comment by joelvbernier** [comment]
I’ll upload it today

**Comment by cryos** [comment]
We ended up discussing this a lot in the Tomviz project where most detector data is also 8 bit or 16 bit unsigned. After a lot of testing, and comparison it was clear that we should never use anything above float32. We rewrote a lot of code to avoid the promotion to float64 which is often a mistake in Python-based pipelines taken out of an abundance of caution. Too much accuracy with no value costs us a lot in memory and processing time. I would think the same is true here, no matter what Python tells us a detector dark field doesn't need more than float32, it may well be reasonable to truncate derived detector data back to 16 bit unsigned integers, but we won't need double precision.

**Comment by joelvbernier** [comment]
I think it makes sense to cast the dark image as the same dtype as the input.

**Comment by psavery** [comment]
I think casting it back to the original dtype sounds good to me, but either is good.

@donald-e-boyce Is there an easy way to convert the dtype of an imageseries?

**Comment by joelvbernier** [comment]
Hmmm, in theory it looks like the `_toarray()` helper function in `imageseries.stats` is supposed to be casting back to the dtype attribute set on the input.  @donald-e-boyce?

**Comment by donald-e-boyce** [comment]
_toarray() is only used in percentile/median, but even there, they are only
used as input to the numpy function, which usually returns float. It would
be a simple fix to convert the output of percentile/median to the original
type.Min/max are computed frame by frame in a way that doesn't change the
dtype, but average() does do a division. In any case, this would be easy.
Should I do this?

**Comment by joelvbernier** [comment]
I think preserving the dtype is probably the best practice, or at worst int32.


On May 9, 2020 at 8:39 AM, Donald Boyce <notifications@github.com> wrote:






_toarray() is only used in percentile/median, but even there, they are only
used as input to the numpy function, which usually returns float. It would
be a simple fix to convert the output of percentile/median to the original
type.Min/max are computed frame by frame in a way that doesn't change the
dtype, but average() does do a division. In any case, this would be easy.
Should I do this?

**Comment by joelvbernier** [comment]
@psavery : have you taken a look at the 8-detector case in the examples?

**Comment by psavery** [comment]
Yep, I'm actually running timers now. Will give the results soon...

**Comment by psavery** [comment]
Here are my timing results for a single run each, for the 8-detector case. Looks like the use of float32 sped it up a lot. multiprocess and multithread were about the same speed, with just a little speedup. I'm happy to put up the code soon to see if we can think of improvements.
```
float64: 200.92 s
float32: 124.67 s
multiprocess: 109.24 s
multithread: 107.52 s
```

**Comment by psavery** [comment]
This got fixed by #319.
